x/leverage/keeper/keeper.go

```diff
@@ -96,42 +105,65 @@ func (k Keeper) LendAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, loan sdk.C
 // for the original tokens lent. If the uToken type is invalid or account balance
 // insufficient on either side, we return an error.
 func (k Keeper) WithdrawAsset(ctx sdk.Context, lenderAddr sdk.AccAddress, uToken sdk.Coin) error {
-	if !uToken.IsValid() {
+	if !k.IsAcceptedUToken(ctx, uToken.Denom) {
 		return sdkerrors.Wrap(types.ErrInvalidAsset, uToken.String())
 	}
 
-	// TODO: Calculate lender's borrow limit and current borrowed value, if any.
-	// Prevent withdrawing assets when it would bring user borrow limit below
-	// current borrowed value.
-	//
-	// ref: https://github.com/umee-network/umee/issues/213
-
-	withdrawal, err := k.ExchangeUToken(ctx, uToken)
+	// calculate base asset amount to withdraw
+	token, err := k.ExchangeUToken(ctx, uToken)
 	if err != nil {
 		return err
 	}
 
 	// Ensure module account has sufficient unreserved tokens to withdraw
-	reservedAmount := k.GetReserveAmount(ctx, withdrawal.Denom)
-	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), withdrawal.Denom).Amount
-	if withdrawal.Amount.GT(availableAmount.Sub(reservedAmount)) {
-		return sdkerrors.Wrap(types.ErrLendingPoolInsufficient, withdrawal.String())
-	}
-
-	// send the uTokens from the lender to the module account
-	uTokens := sdk.NewCoins(uToken)
+	reservedAmount := k.GetReserveAmount(ctx, token.Denom)
+	availableAmount := k.bankKeeper.GetBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token.Denom).Amount
+	if token.Amount.GT(availableAmount.Sub(reservedAmount)) {
+		return sdkerrors.Wrap(types.ErrLendingPoolInsufficient, token.String())
+	}
+
+	// Withdraw will first attempt to use any uTokens in the lender's wallet
+	fromWallet := sdk.MinInt(k.bankKeeper.GetBalance(ctx, lenderAddr, uToken.Denom).Amount, uToken.Amount)
```

Comment:
consider a different variable name here 👍

Reply:
now `amountFromWallet` and `amountFromCollateral`